Change LTC68042cell_nextVoltages() to avoid possible stale data, take 2

RetiredBill opened the pull request:
Changed LTC68042cell_nextVoltages() to wait for acquisition to complete.

If LTC68042cell_nextVoltages() is called in GATHER state while an acquisition is in process, it now returns without gathering data and without advancing its internal state, avoiding gathering stale data. The new "WAITING_FOR_ADC" state will introduce an extra event loop iteration in key-on mode.  The next PR will optimize out this extra delay.

RetiredBill committed:
Changed LTC68042cell_nextVoltages() to wait for acqusition to complete.

If LTC68042cell_nextVoltages() is called in GATHER state while an
  acquisition is in process, it now waits for the acquisition to complete
  before getting data (avoiding stale data). This wait should not
  be triggered in key-on mode since calls occur at intervals longer
  than the acquisition time.

RetiredBill committed:
Removed double calls of LTC68042cell_nextVoltages().

Now that LTC68042cell_nextVoltages() waits for an acquisition to
  complete, double calls should not be needed to prevent stale data.

RetiredBill committed:
Simpler implimentation of waiting for acqusition to complete.

Rather than a blocking delay, LTC68042cell_nextVoltages() now just
  defers advancing its internal state until acquisition if complete.

Also reverted the Removed double calls of LTC68042cell_nextVoltages()
  for the moment.

doppelhub commented on Firmware/firmwareLiBCM/src/LTC68042cell.cpp:
```diff
@@ -239,48 +241,56 @@ void processAllCellVoltages(void)
-                chipAddress = FIRST_IC_ADDR; //reset to first LTC IC
-                presentState = LTC_STATE_PROCESS; //all cell voltages gathered.  Process data on next run.
+        // but don't gather data or advance the state if the  current conversion is not complete (should not usually be necessary)
+        if ( ( ! conversionInProcess) || (LTC6804_MAX_CONVERSION_TIME_ms < (millis() - conversionStart_ms)) )
```
Per my other comment, this code will likely go away... but in general I prefer the following formatting:
-explicit boolean comparisons
-newline for each logical compare

```suggestion
        if ( (conversionInProcess == false)                                    ||
             (LTC6804_MAX_CONVERSION_TIME_ms < (millis() - conversionStart_ms)) )
```

RetiredBill replied:
Fixed in next iteration

doppelhub commented on Firmware/firmwareLiBCM/src/LTC68042cell.cpp:
```diff
@@ -239,48 +241,56 @@ void processAllCellVoltages(void)
 bool LTC68042cell_nextVoltages(void)
 {
     static uint8_t presentState = LTC_STATE_FIRSTRUN;
+    static bool conversionInProcess = false; //used to speed up execution of conversion complete test
```
Why not just add a new state to the existing 'presentState' state machine?  For example, 'LTC_STATE_GATHER' stores millis() in conversionStart_ms and then sets presentState = LTC_WAITING_FOR_ADC_CONVERSION, which only passes the state handler to 'LTC_STATE_GATHER' after period LTC6804_MAX_CONVERSION_TIME_ms.

With the above implementation, conversionInProcess isn't necessary, because we're only checking timestamp while waiting for the conversion to finish.

RetiredBill replied:
Fixed in next iteration

doppelhub reviewed (changes requested):
I prefer adding a new state to the existing state machine, versus adding two new variable conditionals to the entire existing machine.  See comments for more details.

RetiredBill committed:
Backported WAITING_FOR_ADC code from WGC_pushTriggeredMode

RetiredBill committed:
Partially reverted for a simpler (less optimized) implementation.

Reverted moving GATHER block into doCellDataGather().
This makes the change simpler, but is less optimum: Now, in key-on mode,
an extra 10ms loop iteration is required to traverse the “WAITING_FOR_ADC”
state.  This will reversion will be undone in the next commit.

Also reverted some style only changes.

RetiredBill commented on Firmware/firmwareLiBCM/src/LTC68042cell.cpp:
```diff
@@ -11,6 +11,7 @@
 //  Example: cellVoltages_counts[0][ 1] is IC_1 cell_02
 //  Example: cellVoltages_counts[3][11] is IC_4 cell_12
 uint16_t cellVoltages_counts[TOTAL_IC][CELLS_PER_IC];
+uint32_t conversionStart_us = 0;
```
I changed to tracking usec because I wanted better resolution for tracking acquisition time, especially with the test code I'll be contributing later.

RetiredBill commented on Firmware/firmwareLiBCM/src/key.cpp:
```diff
@@ -111,16 +111,14 @@ void keyOn_coldBootTasks(void)
     LED(3,ON);
 
-    //process cell voltages
-    while(millis() - timeSinceLTC6804conversionStarted_us < LTC6804_MAX_CONVERSION_TIME_ms) { ; } //wait for conversion to finish
```
This is now redundant, so I removed it.
